### PR TITLE
Implement tokenisation for security code component

### PR DESCRIPTION
**Pull request by okhan-okbay-cko**
## Issue

JIRA-2139

## Proposed changes

Implement tokenisation capability for the security code component.

## Test Steps

> 1. Go to the main screen
> 2. Tap on `Security Code Component` button
> 3. Fill in the fields and tap `Pay Now` button. See the token is received.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [x] Reviewers assigned
* [x] I have performed a self-review of my code and manual testing
* [x] Lint and unit tests pass locally with my changes
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] I have added necessary documentation (if applicable)

### CODE REVIEW

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+//  Created by Okhan Okbay on 05/10/2023.
+//
+
+import Foundation
+
+extension TokenisationError {
+  public enum SecurityCodeError: CheckoutError {
+    case cvvTokenInvalid
+    case cvvTokenUsed
+    case cvvTokenExpired
```
Comment (okhan-okbay-cko) [inline Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift:14]: which error codes shall we have here? @ehab-al-cko @deepesh-vasthimal-cko

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 2]
```diff
@@ -170,3 +171,53 @@ final public class CheckoutAPIService: CheckoutAPIProtocol {
...
+//        publicKey: publicKey
+//      )))
+      createSecurityCodeToken(requestParameters: requestParameters, completion: completion)
+    case .failure(let error):
+      switch error {
+      case .baseURLCouldNotBeConvertedToComponents, .couldNotBuildURL:
+        completion(.failure(.couldNotBuildURLForRequest))
+      }
+    }
+  }
```
Comment (chintan-soni-cko) [inline Checkout/Source/Tokenisation/CheckoutAPIService.swift:200]: Before hitting the tokenisation API endpoint, do we validate the CVV SDK side and throw an error message to the user if they
1. Hit the pay button without adding CVV
2. If they inject AMEX and hit the pay button with 3 characters, should throw SDK-side internal error
3. If they inject no scheme (in this case unknown), hitting the pay button with less than 3 characters should throw SDK-side internal error

Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:200]: 1) returns false if a scheme is passed
2) yes, it returns false when the scheme is AMEX and there are 3 digits typed in
3) I'll need to add an extra check for the 0 digit case when there is no scheme. Currently it accepts `0, 3 and 4 digits`

@chintan-soni-cko thanks for the detailed review 🙂

Comment (chintan-soni-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:200]: Thanks, Okhan. Sorry if misunderstood your answers. Are you referring here isEnteredCVVValid value or when hitting the tokenisation endpoint? In Android, there are no flags on whether it is valid or not while hitting the tokenisation. Just returning failure callback with an error message
Screenshot 2023-10-11 at 15 34 27

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift (hunk 1 above)
Comment (chintan-soni-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift:14]: @okhan-okbay-cko there are no static error codes. We return directly what we are receiving from the tokenisation end point
You can test it in postman
curl --location 'https://api.sandbox.checkout.com/tokens' \
--header 'Content-Type: application/json' \
--header 'Authorization: pk_6b30805a-1f3b-4c63-8b75-eb3030109173' \
--header 'User-Agent: checkout-android/1.0.0' \
--data '{
  "type": "cvv",
  "token_data": {
    "cvv": ""
  }
}'

File: Checkout/Source/Model/SecurityCodeTokenDetails.swift [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// 
+//  File.swift
+//  
+//
+//  Created by Okhan Okbay on 10/10/2023.
+//
+
+import Foundation
+
+public struct SecurityCodeTokenDetails {
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Model/SecurityCodeTokenDetails.swift:10]: This is called `CVVTokenDetails` in the case of Android. We should aim to keep things common whichever we can and if possible.

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// 
+//  SecurityCodeResponse.swift
+//  
+//
+//  Created by Okhan Okbay on 05/10/2023.
+//
+
+import Foundation
+
+struct SecurityCodeResponse: Decodable {
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:10]: Is there a reason to have SecurityCodeResponse and SecurityCodeTokenDetails which are similar data models?

File: Checkout/Source/Network/RequestFactory.swift [hunk 5]
```diff
@@ -37,18 +37,21 @@ final class RequestFactory: RequestProviding {
 
   enum Request: Equatable {
     case token(tokenRequest: TokenRequest, publicKey: String)
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Network/RequestFactory.swift:39]: Not related to PR but should we call the `token` as `cardToken`?

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 6]
```diff
@@ -170,3 +171,53 @@ final public class CheckoutAPIService: CheckoutAPIProtocol {
     }
   }
 }
+
+// MARK: Security Code Request
+
+extension CheckoutAPIService {
+  public func createSecurityCodeToken(securityCode: String, completion: @escaping (Result<SecurityCodeTokenDetails, TokenisationError.SecurityCodeError>) -> Void) {
+    guard !publicKey.isEmpty else {
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Tokenisation/CheckoutAPIService.swift:179]: How is the publicKey being injected in case of CVV token creation?

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 7]
```diff
@@ -170,3 +171,53 @@ final public class CheckoutAPIService: CheckoutAPIProtocol {
...
+        completion(.failure(.missingAPIKey))
+        return
+    }
+
+    let request = SecurityCodeRequest(tokenData: TokenData(securityCode: securityCode))
+    let requestParameterResult = requestFactory.create(request: .securityCodeToken(request: request, publicKey: publicKey))
+
+    switch requestParameterResult {
+    case .success(let requestParameters):
+//      logManager.queue(event: .tokenRequested(CheckoutLogEvent.TokenRequestData(
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Tokenisation/CheckoutAPIService.swift:189]: Is this intentionally commented ?

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 8]
```diff
@@ -170,3 +171,53 @@ final public class CheckoutAPIService: CheckoutAPIProtocol {
...
+    }
+  }
+
+  private func createSecurityCodeToken(requestParameters: NetworkManager.RequestParameters, completion: @escaping (Result<SecurityCodeTokenDetails, TokenisationError.SecurityCodeError>) -> Void) {
+    requestExecutor.execute(
+      requestParameters,
+      responseType: SecurityCodeResponse.self,
+      responseErrorType: TokenisationError.ServerError.self
+    ) { [tokenDetailsFactory, logManager, logTokenResponse] tokenResponseResult, httpURLResponse in
+//      logTokenResponse(tokenResponseResult, httpURLResponse)
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Tokenisation/CheckoutAPIService.swift:208]: Is this intentionally commented ?

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 6 above)
Comment (deepesh-vasthimal-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:179]: Please ignore I can see the public key is injected in `SecurityCodeComponent`.

File: iOS Example Frame/iOS Example Frame/Factory/Factory.swift [hunk 9]
```diff
@@ -16,6 +16,7 @@ enum Factory {
     // swiftlint:disable:next force_unwrapping
   static let failureURL = URL(string: "https://httpstat.us/403")!
   static let apiKey = "pk_sbox_ym4kqv5lzvjni7utqbliqs2vhqc"
+  static let securityCodeAPIKey = "pk_6b30805a-1f3b-4c63-8b75-eb3030109173"
```
Comment (deepesh-vasthimal-cko) [inline iOS Example Frame/iOS Example Frame/Factory/Factory.swift:19]: Ideally, this should be just one apiKey or we are using this new key because we don't have our own CC @chintan-soni-cko. If we don't have our own test key we should enable the existing test key to support the CVV token as required.

**Review by deepesh-vasthimal-cko — CHANGES_REQUESTED** [review]
Few minor comments otherwise looks good.

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 2 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:200]: correct. it will be the same for iOS as in your screenshot @chintan-soni-cko

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift (hunk 1 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift:14]: It's an error that can be thrown from the `RequestFactory.swift` @deepesh-vasthimal-cko

Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift:14]: Are these error code just randomly defined by us for logging purposes? @ehab-al-cko @deepesh-vasthimal-cko This question is not relevant to the CVV component and overall a Frames question

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift (hunk 4 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:10]: `SecurityCodeResponse` is an internal `Decodable` model which we can change anytime by our needs. `SecurityCodeTokenDetails` is a public frozen model we expose to merchants

Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:10]: Not a solid reason since they are the exact same models property-wise. But can be good to keep them separate. A bit or premature optimisation maybe but it doesn't hurt that much to keep them I believe 🙂

File: Checkout/Source/Network/RequestFactory.swift (hunk 5 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/RequestFactory.swift:39]: good shout, will do since it's an internal type

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 7 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:189]: yeah, to mark where we will log it and how and make me remember during the logging ticket work

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 8 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:208]: yeah. will be done during the logging ticket.

File: iOS Example Frame/iOS Example Frame/Factory/Factory.swift (hunk 9 above)
Comment (okhan-okbay-cko) [reply iOS Example Frame/iOS Example Frame/Factory/Factory.swift:19]: As discussed offline, this is being worked on and will be updated.

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift (hunk 1 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift:14]: discussed this offline. on hold till next visit

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 2 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:200]: as discussed offline, Android handles and returns error differently. On Android, there are error messages being returned, whereas in iOS we return error codes and the merchants' developers decide what to show as an error message. We'll keep the same approach.

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift (hunk 1 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeError.swift]: updated the error codes now. Following the same style of what iOS has been doing so far and returning an error code back to the merchants' developers. They'll need to assign the relevant error messages to these error codes as it's wanted to be done on Android.

File: Checkout/Source/Network/RequestFactory.swift (hunk 5 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/RequestFactory.swift]: updated this to `cardToken`. We need no doc changes since it's an internal type and this is not a breaking change thereof.

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 7 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift]: removed this line

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift (hunk 4 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:10]: resolving this. Feel free to re-open if doesn't feel like been justified. @deepesh-vasthimal-cko

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 10]
```diff
@@ -170,3 +171,44 @@ final public class CheckoutAPIService: CheckoutAPIProtocol {
     }
   }
 }
+
+// MARK: Security Code Request
+
+extension CheckoutAPIService {
+  public func createSecurityCodeToken(securityCode: String, completion: @escaping (Result<SecurityCodeResponse, TokenisationError.SecurityCodeError>) -> Void) {
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Tokenisation/CheckoutAPIService.swift:178]: For public functions, Just wondering do you need to add any comments for documentation ?

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 11]
```diff
@@ -170,3 +171,44 @@ final public class CheckoutAPIService: CheckoutAPIProtocol {
...
+      createSecurityCodeToken(requestParameters: requestParameters, completion: completion)
+    case .failure(let error):
+      switch error {
+      case .baseURLCouldNotBeConvertedToComponents, .couldNotBuildURL:
+        completion(.failure(.couldNotBuildURLForRequest))
+      }
+    }
+  }
+
+  private func createSecurityCodeToken(requestParameters: NetworkManager.RequestParameters, completion: @escaping (Result<SecurityCodeResponse, TokenisationError.SecurityCodeError>) -> Void) {
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Source/Tokenisation/CheckoutAPIService.swift:198]: Is there something wrong with the formatting here ?

File: Source/Core/CheckoutAPIService.swift [hunk 12]
```diff
@@ -14,6 +14,7 @@ protocol CheckoutAPIProtocol {
     var logger: FramesEventLogging { get }
     init(publicKey: String, environment: Environment)
     func createToken(_ paymentSource: PaymentSource, completion: @escaping (Result<TokenDetails, TokenisationError.TokenRequest>) -> Void)
+    func createSecurityCodeToken(securityCode: String, completion: @escaping (Result<SecurityCodeTokenDetails, TokenisationError.SecurityCodeError>) -> Void)
```
Comment (deepesh-vasthimal-cko) [inline Source/Core/CheckoutAPIService.swift:17]: formatting looks a bit odd here as well and on line 40 below. I will wait for you to confirm before I add more commets which are similar.

File: Source/Core/CheckoutAPIService.swift [hunk 13]
```diff
@@ -35,4 +36,15 @@ public final class CheckoutAPIService: CheckoutAPIProtocol {
     public func createToken(_ paymentSource: PaymentSource, completion: @escaping (Result<TokenDetails, TokenisationError.TokenRequest>) -> Void) {
         checkoutAPIService.createToken(paymentSource, completion: completion)
     }
+
+    public func createSecurityCodeToken(securityCode: String, completion: @escaping (Result<SecurityCodeTokenDetails, TokenisationError.SecurityCodeError>) -> Void) {
```
Comment (deepesh-vasthimal-cko) [inline Source/Core/CheckoutAPIService.swift:40]: For public functions can we add comments or documentation?

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift [hunk 14]
```diff
@@ -9,16 +9,26 @@ import Checkout
...
+  var view: SecurityCodeView!
+  
+  // func configure(_:_:) arguments
+  var configuration: SecurityCodeComponentConfiguration!
+  var isSecurityCodeValid: ((Bool) -> Void)!
 
-  private var configuration: SecurityCodeComponentConfiguration!
-  private var isSecurityCodeValid: ((Bool) -> Void)!
-  private var cardValidator: CardValidating!
+  // func configure(_:_:) initialised properties. They depend on the configuration argument
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:19]: What is the purpose of this comment?

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift [hunk 15]
```diff
@@ -35,23 +45,25 @@ extension SecurityCodeComponent {
                         isSecurityCodeValid: @escaping (Bool) -> Void) {
     self.configuration = configuration
     self.isSecurityCodeValid = isSecurityCodeValid
-
-    self.cardValidator = CardValidator(environment: configuration.environment.checkoutEnvironment)
-
+    
+    cardValidator = CardValidator(environment: configuration.environment.checkoutEnvironment)
+    checkoutAPIService = CheckoutAPIService(publicKey: configuration.apiKey,
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:50]: If I understand this right, The SecurityCodeComponent.swift requires a checkoutAPIService and cardValidator. And if merchants were to create a card token then there would also be another CheckoutAPIService instance created by the merchant app using the API key and environment. Also here we are passing the CardValidator which requires environment. Not a blocker for this PR but wanted to understand, Is there a way to improve this? 
Also

**Review by deepesh-vasthimal-cko — DISMISSED** [review]
LGTM few minor comments not blockers.

**Comment by deepesh-vasthimal-cko** [comment]
Also check for [17 Code Smells](https://sonarcloud.io/project/issues?id=checkout_frames-ios&pullRequest=490&resolved=false&types=CODE_SMELL) if anything critical to fix we should do it now.

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 10 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:178]: This has been discussed offline and we wanted to follow the path of the `createToken` function in the same file since we don't encourage the usage of this function in headless mode but still have to define it as `public` since we need to expose it to `Frames`. (This function lives under `Checkout` which is another module) @deepesh-vasthimal-cko

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 11 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:198]: Can you elaborate please? @deepesh-vasthimal-cko

File: Source/Core/CheckoutAPIService.swift (hunk 12 above)
Comment (okhan-okbay-cko) [reply Source/Core/CheckoutAPIService.swift:17]: can you elaborate on the oddness please? @deepesh-vasthimal-cko

File: Source/Core/CheckoutAPIService.swift (hunk 13 above)
Comment (okhan-okbay-cko) [reply Source/Core/CheckoutAPIService.swift:40]: We don't want to advertise this as a function that we want anyone to use. We want all merchants to use our UI component directly. This is discussed offline under a Slack thread I'll mention you now @deepesh-vasthimal-cko

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift (hunk 14 above)
Comment (okhan-okbay-cko) [reply Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:19]: Juts an internal one. Categorising the properties. Otherwise I tend to leave a blank line in between the different functional groupings of properties and it doesn't always look good without a proper explanation @deepesh-vasthimal-cko so, no harm in adding these.

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift (hunk 15 above)
Comment (okhan-okbay-cko) [reply Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:50]: Definitely a good catch and question. We can have a look if there is a way to use the same instance. 

But also, I'm not so sure of the balance of added-value/effort would be high of that improvement. `CheckoutAPIService` is cheap. Nothing big or memory consuming or complicated. Creating it many times shouldn't be a problem and trying to have just one instance of it might be over optimising the codebase. 

In any case, noted and will have a look. Thanks. 
@deepesh-vasthimal-cko

**Comment by okhan-okbay-cko** [comment]
Code smells are in unit tests and they don't affect the code quality but still we can get rid of them. Will have a look @deepesh-vasthimal-cko 

> Also check for [17 Code Smells](https://sonarcloud.io/project/issues?id=checkout_frames-ios&pullRequest=490&resolved=false&types=CODE_SMELL) if anything critical to fix we should do it now.

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift [hunk 16]
```diff
@@ -35,23 +45,25 @@ extension SecurityCodeComponent {
                         isSecurityCodeValid: @escaping (Bool) -> Void) {
     self.configuration = configuration
     self.isSecurityCodeValid = isSecurityCodeValid
-
-    self.cardValidator = CardValidator(environment: configuration.environment.checkoutEnvironment)
-
+    
+    cardValidator = CardValidator(environment: configuration.environment.checkoutEnvironment)
```
Comment (chintan-soni-cko) [inline Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:49]: I can see cardValidator has all the validators, for Security Component, can we inject only the `CVVValidator` only instead of cardValidator? In Android, we have independently added `CVVValidator` in this way we avoid unnecessary multiple validators for Securitycomponent's scope

File: CheckoutTests/Tokenisation/CheckoutAPIServiceTests.swift [hunk 17]
```diff
@@ -200,3 +205,106 @@ final class CheckoutAPIServiceTests: XCTestCase {
     XCTAssertEqual(subject.correlationID, stubCorrelationID)
   }
 }
+
+extension CheckoutAPIServiceTests {
```
Comment (chintan-soni-cko) [inline CheckoutTests/Tokenisation/CheckoutAPIServiceTests.swift:209]: Can we add a unit test, when `securitycode` is empty or invalid length (for instance, 6 numbers)? In this case, we will get an internal SDK error I assume.

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+//  Created by Okhan Okbay on 05/10/2023.
+//
+
+import Foundation
+
+public struct SecurityCodeResponse: Decodable, Equatable {
+  public let type: String
+  public let token: String
+  public let expiresOn: String
+}
```
[the remark this answers is not in the document]
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:14]: good shout! 

In addition, we can add some sample responses to denote that type means `cvv` string and can only mean that. Also we can imply the format of the expiry date that's returning here. What do you think? @chintan-soni-cko

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift (hunk 16 above)
Comment (okhan-okbay-cko) [reply Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:49]: I'd seen your discussion under the Android repo and tried to do the same to our side. It meant refactoring for the `SecurityCodeViewModel` since it accepts `CardValidating` protocol conformant type at the moment. I didn't want to make that change because it'd introduce more moving pieces to this work and in the end the added value wouldn't worth that much of an update I decided. But thanks for this great shout! @chintan-soni-cko

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (chintan-soni-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:14]: Yes, it would be nice to have definitely. However, in this case, they will check in-depth details from the Checkout public docs as well (once they publish for CVV) because we covered almost all the details for the parameters in the SDK documentation.

[the remark this answers is not in the document]
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift:14]: Thanks for this. The current comments on Android code seems like repeating what the code is already speaking about without the comments, I believe. I'm not saying they are redundant but they don't feel like crucial. I think we can either add more context or can discuss the value of the comments otherwise @chintan-soni-cko

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 10 above)
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:178]: resolving this as I believe the offline discussion decided to keep it this way. feel free to re-open if it doesn't feel right

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift [hunk 19]
```diff
@@ -11,6 +11,7 @@ import CheckoutEventLoggerKit
 
 public protocol CheckoutAPIProtocol {
   func createToken(_ paymentSource: PaymentSource, completion: @escaping (Result<TokenDetails, TokenisationError.TokenRequest>) -> Void)
+  func createSecurityCodeToken(securityCode: String, completion: @escaping (Result<SecurityCodeResponse, TokenisationError.SecurityCodeError>) -> Void)
```
[the remark this answers is not in the document]
Comment (okhan-okbay-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:14]: resolving this as I believe there's nothing to be changed since we comply with the API Spec. Feel free to re-open for any concern.

File: Source/Core/CheckoutAPIService.swift (hunk 12 above)
Comment (okhan-okbay-cko) [reply Source/Core/CheckoutAPIService.swift:17]: Screenshot 2023-10-31 at 12 57 26

it looks like this in my Xcode @deepesh-vasthimal-cko

File: Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift (hunk 14 above)
Comment (okhan-okbay-cko) [reply Source/UI/SecurityCodeComponent/SecurityCodeComponent.swift:19]: resolving this but feel free to re-open for further discussion

File: Checkout/Source/Tokenisation/CheckoutAPIService.swift (hunk 11 above)
Comment (deepesh-vasthimal-cko) [reply Checkout/Source/Tokenisation/CheckoutAPIService.swift:198]: Your screenshot below looks good, It would just be my chrome showing things in next line

File: CheckoutTests/Tokenisation/CheckoutAPIServiceTests.swift (hunk 17 above)
Comment (okhan-okbay-cko) [reply CheckoutTests/Tokenisation/CheckoutAPIServiceTests.swift:209]: I had a look at this and I think for the api service itself, it's a bit of over-responsibility to run a pre-check over the provided security code. So, it's being done in `SecurityCodeComponent`. Added some invalid input tests for 5 digits into `SecurityCodeComponentTests` to cover those scenarios @chintan-soni-cko

File: Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (okhan-okbay-cko) [reply Checkout/Source/Network/Models/SecurityCode/SecurityCodeResponse.swift]: added comments @chintan-soni-cko

**Review by chintan-soni-cko — APPROVED** [review]
Screenshot 2023-11-01 at 15 42 01
LGTM. Just one question. In Android we are providing error message ( for instance, "Please enter valid cvv code" incase of invalid CVV) and we have error codes in IOS which is fine. However, in IOS we have `localizedDescription` from error. In this case as attached screenshot for invalid CVV. Can we customise it to userfriendly message and error code displaying as 4. Should it be 3006?

**Review by chintan-soni-cko — DISMISSED** [review]
